text/0000-improved-binary-gem-support.md

```diff
@@ -0,0 +1,466 @@
+- Feature Name: Improved binary gem support in RubyGems
+- Start Date: 2025-08-20
+- RFC PR: (leave this empty)
+- Bundler Issue: (leave this empty)
+
+# Summary
+
+Add support to RubyGems for more fine-grained matching for binary gem distribution through ruby/api tags and platform tags (`gb-{abi_tags}-{platform_tags}`). This enables "platform matching" based on Ruby ABI version, os, os version, cpu architecture, and libc implementation (for generic linux platforms).
+
+# Motivation
+
+Binary gem distribution currently uses coarse-grained platform matching (e.g., "x86_64-linux") that cannot distinguish between:
+
+- Different Ruby language and ABI versions (Ruby 3.2 vs 3.3 native extension binaries are incompatible)
+- Linux libc implementation versions (musl and glibc can introduce new symbols in new versions)
+
+This leads to native extension gems either failing at runtime, requiring compilation from source even when compatible binaries exist, or packaging _multiple_ extension binaries in a single gem (of which each installation will use only one). This proposal is heavily inspired by the [`wheel` format in the Python ecosystem](https://packaging.python.org/en/latest/specifications/platform-compatibility-tags/) that addresses similar issues.
+
+# Guide-level explanation
+
+When installing gems with native extensions, RubyGems will now automatically select the most appropriate pre-compiled binary based on your exact Ruby version and platform:
+
+```bash
+$ gem install nokogiri
+# On CRuby 3.3, x86_64 Linux with glibc 2.28:
+# Installs nokogiri-1.16.0-gb-cr33-glibclinux_2_28_x86_64.gem (pre-compiled)
+
+# On Ruby 3.2, same platform:
+# Falls back to nokogiri-1.16.0.gem (compiles from source)
+```
+
+Gem authors can build platform-specific gems:
+
+```sh
+gem build --platform "gb-cr33-x86_64_linux"  # Specific to CRuby 3.3 on Linux x86_64
+```
+
+
+The new binary gem platform format follows the pattern `gb-{abi_tags}-{platform_tags}`:
+
+- `abi_tags`: Ruby ABI version (e.g., `rb33` for Ruby 3.3, `any` for pure Ruby, `cr33` for CRuby 3.3, `jr92` for JRuby 9.2, `tr241` for TruffleRuby 24.1)
```

Comment:
We shouldn't transform the ABI version, we should take it as-is from `RbConfig::CONFIG['ruby_version']`.
As said in https://github.com/rubygems/rfcs/pull/60#discussion_r2309136883 the ABI version may evolve in non-predictable ways. Only exact match of `RbConfig::CONFIG['ruby_version']` are compatible.
Another problematic example with e.g. `jr92` is then `jr10` could mean JRuby 1.0 or 10.0.

Reply:
Regarding shortening the RUBY_ENGINE to two letters, I think that's fine (shouldn't cause issues), and I think it's good to use `cr33` instead of `ruby33` for clarity.